### PR TITLE
fix #4635 provider throw RpcException custom retry not correct times

### LINKED ISSUE #4635: when provider throw RpcException consumer do not retry correct times
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [x] I have checked the [FAQ](https://github.com/apache/dubbo/blob/master/FAQ.md) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version: 2.7.1
* Operating System version: windows10
* Java version: jdk1.8

### Steps to reproduce this issue

1. consumer set retries 5
2. connection over executes on provider
3. provider throw RpcException

### Expected Result

consumer will retry 5 times

### Actual Result

when provider throw RpcException，FailoverClusterInvoker  receive a AppReponse with a RpcException in it，and FailoverClusterInvoker can not catch the RpcException and do not retry；

### PR DESCRIPTION
**Pull request by codeimport**
## What is the purpose of the change

fix #4635  provider throw RpcException custom retry not correct times

## Brief changelog

FailoverClusterInvoker;
FailoverRpcExceptionTest;

## Verifying this change

org.apache.dubbo.rpc.cluster.support.FailoverRpcExceptionTest;

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java [hunk 1]
```diff
@@ -91,6 +91,11 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
                             + " using the dubbo version " + Version.getVersion() + ". Last error is: "
                             + le.getMessage(), le);
                 }
+
+                if(result != null && result.getException() != null && result.getException() instanceof RpcException){
```
[the remark this answers is not in the document]
Comment (codeimport) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java]: yet ,i will use result.hasException，but here result may be null，so result also judged null

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java [hunk 2]
```diff
@@ -91,6 +91,11 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
                             + " using the dubbo version " + Version.getVersion() + ". Last error is: "
                             + le.getMessage(), le);
                 }
+
+                if(result != null && result.getException() != null && result.getException() instanceof RpcException){
+                    throw result.getException();
```
[the remark this answers is not in the document]
Comment (codeimport) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java]: i think that here only can throw RpcException ，can not throw biz exception，or consumer will not invoke mock

File: dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailoverRpcExceptionTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+
+/**
+ * @description:
+ * @author: chengang6
```
[the remark this answers is not in the document]
Comment (codeimport) [reply dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailoverRpcExceptionTest.java]: yet ,i will remove author

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (beiwei30) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java]: do you mean RpcException is explicitly thrown from business code on the provider side?

[the remark this answers is not in the document]
Comment (codeimport) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailoverClusterInvoker.java]: the RpcException throw by Dubbo self filter not business code，eg：
when the connection over executes on provider side , the RpcException will throw in ExecuteLimitFilter

File: dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailoverRpcExceptionTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.rpc.cluster.support;
+
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.rpc.*;
```
Comment (AlbumenJ) [inline dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailoverRpcExceptionTest.java:20]: pls do not use comma import

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Pull request closed without merging**